core/src/services/b2/backend.rs

```diff
@@ -0,0 +1,553 @@
+                write_multi_max_size: if cfg!(target_pointer_width = "64") {
+                    Some(5 * 1024 * 1024 * 1024)
+                } else {
+                    Some(usize::MAX)
+                },
+
+                create_dir: true,
+                delete: true,
+                copy: true,
+
+                list: true,
+                list_with_limit: true,
+                list_with_start_after: true,
+                list_with_recursive: true,
+                list_without_recursive: true,
+
+                presign: true,
+                presign_read: true,
+                presign_write: true,
+                presign_stat: true,
+
+                ..Default::default()
+            });
+
+        am
+    }
+
+    async fn read(&self, path: &str, _args: OpRead) -> Result<(RpRead, Self::Reader)> {
+        let resp = self.core.download_file_by_name(path).await?;
+
+        let status = resp.status();
+
+        match status {
+            StatusCode::OK | StatusCode::PARTIAL_CONTENT => {
+                let size = parse_content_length(resp.headers())?;
+                Ok((RpRead::new().with_size(size), resp.into_body()))
+            }
+            StatusCode::RANGE_NOT_SATISFIABLE => Ok((RpRead::new(), IncomingAsyncBody::empty())),
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    async fn create_dir(&self, path: &str, _: OpCreateDir) -> Result<RpCreateDir> {
+        let resp: http::Response<IncomingAsyncBody> = self
+            .core
+            .upload_file(path, Some(0), &OpWrite::default(), AsyncBody::Empty)
+            .await?;
+
+        let status = resp.status();
+
+        match status {
+            StatusCode::OK => {
+                resp.into_body().consume().await?;
+                Ok(RpCreateDir::default())
+            }
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    async fn write(&self, path: &str, args: OpWrite) -> Result<(RpWrite, Self::Writer)> {
+        let writer = B2Writer::new(self.core.clone(), path, args);
+
+        let w = oio::MultipartUploadWriter::new(writer);
+
+        Ok((RpWrite::default(), w))
+    }
+
+    async fn stat(&self, path: &str, _args: OpStat) -> Result<RpStat> {
+        let delimiter = if path.ends_with('/') { Some("/") } else { None };
+        let resp = self
+            .core
+            .list_file_names(Some(path), delimiter, None, None)
+            .await?;
+
+        let status = resp.status();
+
+        match status {
+            StatusCode::OK => {
+                let bs = resp.into_body().bytes().await?;
+
+                let resp: ListFileNamesResponse =
+                    serde_json::from_slice(&bs).map_err(new_json_deserialize_error)?;
+                if resp.files.is_empty() {
+                    return Err(Error::new(ErrorKind::NotFound, "no such file or directory"));
+                }
+                let meta: Metadata = (&resp.files[0]).into();
```

Comment:
Please don't implement `From<Xxx> for Metadata` which could leak our internal struct types. Implement a `parse_file_info` instead.

Reply:
fixed